src/js/node/fs.promises.ts

```diff
@@ -38,32 +38,51 @@ export function watch(
   if (typeof options === "string") {
     options = { encoding: options };
   }
-  fs.watch(filename, options || {}, (eventType: string, filename: string | Buffer | undefined) => {
+  const watcher = fs.watch(filename, options || {}, (eventType: string, filename: string | Buffer | undefined) => {
     events.push({ eventType, filename });
     if (nextEventResolve) {
       const resolve = nextEventResolve;
       nextEventResolve = null;
       resolve();
     }
   });
+
   return {
-    async *[Symbol.asyncIterator]() {
+    [Symbol.asyncIterator]() {
       let closed = false;
-      while (!closed) {
-        while (events.length) {
-          let event = events.shift() as Event;
-          if (event.eventType === "close") {
-            closed = true;
-            break;
+      return {
+        async next() {
+          while (!closed) {
+            while (events.length) {
+              let event = events.shift() as Event;
+              if (event.eventType === "close") {
+                closed = true;
+                break;
+              }
+              if (event.eventType === "error") {
+                closed = true;
+                throw event.filename;
+              }
+              return { value: event, done: false };
+            }
+            await new Promise((resolve: Function) => (nextEventResolve = resolve));
```

Comment:
This can be `Promise.withResolvers` and that will skip creating the extra temporary function

Reply:
fix on https://github.com/oven-sh/bun/pull/3846/commits/897082a4fddabe926c3a5a6661709fc45c8ebd80